lib/sitemap/generator.rb

```diff
@@ -0,0 +1,150 @@
+      @search_client    = Services.elasticsearch(hosts: search_config.base_uri, timeout: 10)
+      @uploader         = uploader
+      @timestamp        = timestamp
+    end
+
+    def run
+      create_sitemap_index(create_sitemaps(get_all_documents))
+    end
+
+    def create_sitemaps(enumerator)
+      enumerator.with_index.map do |documents, index|
+        batch_number = index + 1
+        create_sitemap(documents, batch_number)
+      end
+    end
+
+    def create_sitemap(documents, batch_number)
+      file_name = "sitemap_#{batch_number}.xml"
+      @uploader.upload(
+        file_content: generate_sitemap_xml(documents),
+        file_name:    file_name,
+      )
+      file_name
+    end
+
+    def create_sitemap_index(sitemaps)
+      @uploader.upload(
+        file_content: generate_sitemap_index_xml(sitemaps),
+        file_name:    "sitemap.xml",
+      )
+    end
+
+    # Generate a sitemap which matches the format specified in https://www.sitemaps.org/protocol.html
+    def generate_sitemap_xml(documents)
+      builder = Nokogiri::XML::Builder.new(encoding: "UTF-8") do |xml|
+        xml.urlset(xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9") do
+          documents.each do |document|
+            xml.url {
+              xml.loc document.url
+              xml.lastmod document.last_updated if document.last_updated
+              xml.priority document.priority
+            }
+          end
+        end
+      end
+      builder.to_xml
+    end
+
+    def generate_sitemap_index_xml(sitemap_filenames)
+      builder = Nokogiri::XML::Builder.new(encoding: "UTF-8") do |xml|
+        xml.sitemapindex(xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9") do
+          sitemap_filenames.each do |sitemap_filename|
+            xml.sitemap {
+              xml.loc "#{base_url}/#{SUB_DIRECTORY}/#{sitemap_filename}"
+              xml.lastmod @timestamp.strftime("%FT%T%:z")
+            }
+          end
+        end
+      end
+      builder.to_xml
+    end
+
+  private
+
+    def get_all_documents
+      page                  = initial_scroll
+      scroll_id             = page["_scroll_id"]
+      documents             = [homepage] + page["hits"]["hits"]
+      total_documents_count = page["hits"]["total"]
+      total_page_count      = total_documents_count.fdiv(SCROLL_BATCH_SIZE).ceil
+
+      Enumerator::Lazy.new(0...total_page_count) do |yielder|
+        if documents.count == SITEMAP_LIMIT + 1
+          sitemaps = documents.map do |document|
+            SitemapPresenter.new(document["_source"], property_boost_calculator)
+          end
+          documents.clear
+        end
+
+        more_documents = scroll(scroll_id)
+
+        if more_documents.any?
+          documents.push(*more_documents)
+        else
+          sitemaps = documents.map do |document|
+            SitemapPresenter.new(document["_source"], property_boost_calculator)
+          end
+        end
+
+        yielder << sitemaps if sitemaps
+      end
```

Comment:
I don't really understand this enumerator body, I'd have expected something like:

```ruby
# yield individual documents and have `generate_sitemap_xml` group them into chunks of SITEMAP_LIMIT
Enumerator::Lazy.new(0..total_page_count) do |yielder|
  break if documents.empty?

  documents.map do |document|
    yielder << SitemapPresenter.new(document["_source"], property_boost_calculator)
  end

  documents = scroll(scroll_id)
end
```

Or

```ruby
# yield chunks of size SITEMAP_LIMIT
chunk = []
Enumerator::Lazy.new(0..total_page_count) do |yielder|
  documents.map do |document|
    chunk << SitemapPresenter.new(document["_source"], property_boost_calculator)
    if chunk.size == SITEMAP_LIMIT
      yielder << chunk
      chunk = []
    end
  end

  documents = scroll(scroll_id)

  if documents.empty?
    yielder << chunk
    break
  end
end
```

Also, it looks like the `scroll_id` changes between requests: https://github.com/alphagov/search-api/blob/master/lib/scroll_enumerator.rb#L47-L49

Can't the existing `ScrollEnumerator` class be re-used somehow?

Reply:
Given that `break` doesn't work and there's nothing built in to group into chunks, which would make option 1 nice, this is a candidate:

```ruby
chunk = []
Enumerator::Lazy.new(0..total_page_count) do |yielder|
  next if documents.empty? and chunk.empty?

  documents.map do |document|
    chunk << SitemapPresenter.new(document["_source"], property_boost_calculator)
    if chunk.size == SITEMAP_LIMIT
      yielder << chunk
      chunk = []
    end
  end

  documents = scroll(scroll_id)

  if documents.empty?
    yielder << chunk
    chunk = []
  end
end
```